mllib/src/main/scala/org/apache/spark/mllib/clustering/AffinityPropagation.scala

```diff
@@ -0,0 +1,496 @@
+ * @param lambda lambda parameter used in the messaging iteration loop
+ * @param normalization Indication of performing normalization
+ * @param symmetric Indication of using symmetric similarity input
+ *
+ * @see [[http://en.wikipedia.org/wiki/Affinity_propagation Affinity propagation (Wikipedia)]]
+ */
+@Experimental
+class AffinityPropagation private[clustering] (
+    private var maxIterations: Int,
+    private var lambda: Double,
+    private var normalization: Boolean,
+    private var symmetric: Boolean) extends Serializable {
+
+  import org.apache.spark.mllib.clustering.AffinityPropagation._
+
+  /** Constructs a AP instance with default parameters: {maxIterations: 100, lambda: `0.5`,
+   *    normalization: false, symmetric: true}.
+   */
+  def this() = this(maxIterations = 100, lambda = 0.5, normalization = false, symmetric = true)
+
+  /**
+   * Set maximum number of iterations of the messaging iteration loop
+   */
+  def setMaxIterations(maxIterations: Int): this.type = {
+    this.maxIterations = maxIterations
+    this
+  }
+ 
+  /**
+   * Get maximum number of iterations of the messaging iteration loop
+   */
+  def getMaxIterations: Int = {
+    this.maxIterations
+  }
+ 
+  /**
+   * Set lambda of the messaging iteration loop
+   */
+  def setLambda(lambda: Double): this.type = {
+    this.lambda = lambda
+    this
+  }
+ 
+  /**
+   * Get lambda of the messaging iteration loop
+   */
+  def getLambda(): Double = {
+    this.lambda
+  }
+ 
+  /**
+   * Set whether to do normalization or not
+   */
+  def setNormalization(normalization: Boolean): this.type = {
+    this.normalization = normalization
+    this
+  }
+ 
+  /**
+   * Get whether to do normalization or not
+   */
+  def getNormalization(): Boolean = {
+    this.normalization
+  }
+
+  /**
+   * Set whether the input similarities are symmetric or not.
+   * When symmetric is set to true, we assume that input similarities only contain triangular
+   * matrix. That means, only s,,ij,, is included in the similarities. If both s,,ij,, and
+   * s,,ji,, are given in the similarities, it very possibly causes error.
+   */
+  def setSymmetric(symmetric: Boolean): this.type = {
+    this.symmetric = symmetric
+    this
+  }
+
+  /**
+   * Get whether the input similarities are symmetric or not
+   */
+  def getSymmetric(): Boolean = {
+    this.symmetric
+  }
+
+  /**
+   * Calculate the median value of similarities
+   */
+  private def getMedian(similarities: RDD[(Long, Long, Double)]): Double = {
+    val sorted: RDD[(Long, Double)] = similarities.sortBy(_._3).zipWithIndex().map {
+      case (v, idx) => (idx, v._3)
+    }.persist()
+
+    val count = sorted.count()
+
+    val median: Double =
+      if (count % 2 == 0) {
+        val l = count / 2 - 1
+        val r = l + 1
+        (sorted.lookup(l).head + sorted.lookup(r).head).toDouble / 2
+      } else {
+        sorted.lookup(count / 2).head
```

Comment:
This is going to do a lookup for a fraction because `(count % 2 !== 0)`. Is that how it's supposed to work?